iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -383,58 +479,86 @@ public override async Task<long> UpdateReportedPropertiesAsync(ReportedPropertie
 
         public override async Task OpenAsync(CancellationToken cancellationToken)
         {
-            // If this object has already been disposed, we will throw an exception indicating that.
-            // This is the entry point for interacting with the client and this safety check should be done here.
-            // The current behavior does not support open->close->open
-            if (_isDisposed)
-            {
-                throw new ObjectDisposedException(nameof(RetryDelegatingHandler));
-            }
-
-            if (_isOpen)
-            {
-                return;
-            }
+            if (Logging.IsEnabled)
+                Logging.Enter(this, cancellationToken, nameof(OpenAsync));
 
-            await _handlerSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
             try
             {
-                if (!_isOpen)
+                // If this object has already been disposed, we will throw an exception indicating that.
+                // This is the entry point for interacting with the client and this safety check should be done here.
+                // The current behavior does not support open->close->open
+                if (_isDisposed)
                 {
-                    if (Logging.IsEnabled)
-                        Logging.Info(this, "Opening connection", nameof(OpenAsync));
+                    throw new ObjectDisposedException("IoT client", ClientDisposedMessage);
+                }
 
-                    // This is to ensure that if OpenInternalAsync() fails on retry expiration with a custom retry policy,
-                    // we are returning the corresponding connection status change event => disconnected: retry_expired.
-                    try
-                    {
-                        await OpenInternalAsync(cancellationToken).ConfigureAwait(false);
-                    }
-                    catch (Exception ex) when (!Fx.IsFatal(ex))
-                    {
-                        HandleConnectionStatusExceptions(ex, true);
-                        throw;
-                    }
+                switch (GetClientTransportStatus())
+                {
+                    case ClientTransportStatus.Open:
+                    case ClientTransportStatus.Opening:
+                        return;
 
-                    if (!_isDisposed)
-                    {
-                        _isOpen = true;
+                    case ClientTransportStatus.Closing:
+                        throw new InvalidOperationException($"The client is currently closing. Wait until {nameof(CloseAsync)} completes and then invoke {nameof(OpenAsync)} again.");
```

Comment:
If `Openasync()` is called while the client is closing, throw an InvalidOperationException.

Reply:
Is this documented on the client's OpenAsync call?